### PR TITLE
Fixed N+1 query occurrence in the /api/v1/facility/

### LINKED ISSUE #2256: Optimize the N+1 query occurrence in the `/api/v1/facility/` endpoint to improve performance and reduce database load.
### Feature Request

**Description**
Optimize the N+1 query occurrence in the `/api/v1/facility/` endpoint to improve performance and reduce database load.

- **ViewSet File**: `care/facility/api/viewsets/facility.py`
- **Serializer**: `care/facility/api/serializers/facility.py`

**Suggested Solution**
To address the N+1 query problem, use Django's `select_related` and `prefetch_related` methods:
1. **Use `select_related`**: For single-valued relationships (e.g., foreign keys, one-to-one relationships) to perform SQL joins and fetch related objects in a single query.
2. **Use `prefetch_related`**: For multi-valued relationships (e.g., many-to-many, reverse foreign keys) to fetch related objects in separate queries and join them in Python.

**Additional Context**
Optimizing these queries should enhance the endpoint's performance, reduce the number of database queries, and lower the overall load on the database.

### PR DESCRIPTION
**Pull request by dhruv-goyal-10**
## Changes Made

Annotated the count of beds & patients in Facility Queryset, which were causing N + 1 query problem.

### Associated Issue

- Fixes #2256 

## Merge Checklist

- [X] Tests added/fixed
- [ ] Update docs in `/docs`
- [X] Linting Complete
- [ ] Any other necessary step

@coronasafe/care-backend-maintainers @coronasafe/care-backend-admins

### CODE REVIEW

**Comment by dhruv-goyal-10** [comment]
@gigincg Can you have a look at the PR?

Thanks

**Comment by gigincg** [comment]
@sainak Can you get this reviewed

cc: @dhruv-goyal-10

**Review by sainak — COMMENTED** [review]
LGTM, can you add a test case for the list api to check the number of queries 
like this:
https://github.com/coronasafe/care/blob/develop/care/facility/tests/test_bed_api.py#L38

**Comment by gigincg** [comment]
@dhruv-goyal-10 

> LGTM, can you add a test case for the list api to check the number of queries like this: https://github.com/coronasafe/care/blob/develop/care/facility/tests/test_bed_api.py#L38

**Comment by dhruv-goyal-10** [comment]
@sainak @gigincg I have added the test case for the same.

File: care/facility/tests/test_facility_api.py [hunk 1]
```diff
@@ -18,7 +18,8 @@ def test_all_listing(self):
         self.assertIs(response.status_code, status.HTTP_200_OK)
 
     def test_listing(self):
```
Comment (sainak) [inline care/facility/tests/test_facility_api.py:24]: create a few more facilities and some beds and patients

**Comment by dhruv-goyal-10** [comment]
@sainak Have a look at the PR now...

**Review by sainak — APPROVED** [review]

File: care/facility/api/viewsets/facility.py [hunk 2]
```diff
@@ -71,8 +72,15 @@ class FacilityViewSet(
...
     """Viewset for facility CRUD operations."""
 
-    queryset = Facility.objects.all().select_related(
-        "ward", "local_body", "district", "state"
+    queryset = (
+        Facility.objects.all()
+        .select_related("ward", "local_body", "district", "state")
+        .annotate(
+            bed_count=Count("bed"),
+            patient_count=Count(
```
Comment (rithviknishad) [inline care/facility/api/viewsets/facility.py:80]: Shouldn't we exclude related beds and patients that are deleted?

Comment (rithviknishad) [reply care/facility/api/viewsets/facility.py:80]: cc: @sainak

**Comment by rithviknishad** [comment]
Closing as new issue has been opened with a more appropriate solution: https://github.com/ohcnetwork/care/issues/2481

**Pull request closed without merging**